### PR TITLE
672 add mypy pre commit hook

### LINKED ISSUE #672: Add mypy pre-commit hook to regulate type hints
## Subject of the feature

We should add https://github.com/python/mypy to uniform the type hinting in py code.

### PR DESCRIPTION
**Pull request by mathpluscode**
# Description

- update some pre-commit hooks
- add mypy and fix all type hint errors

Fixes #672

## Type of change

What types of changes does your code introduce to DeepReg?

_Please check the boxes that apply after submitting the pull request._

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation Update (fix or improvement on the documentation)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Other (if none of the other choices apply)

## Checklist

_Please check the boxes that apply after submitting the pull request._

_If you're unsure about any of them, don't hesitate to ask. We're here to help! This is
simply a reminder of what we are going to look for before merging your code._

- [ ] I have
      [installed pre-commit](https://deepreg.readthedocs.io/en/latest/contributing/setup.html)
      using `pre-commit install` and formatted all changed files. If you are not
      certain, run `pre-commit run --all-files`.
- [ ] My commits' message styles matches
      [our requested structure](https://deepreg.readthedocs.io/en/latest/contributing/commit.html),
      e.g. `Issue #<issue number>: detailed message`.
- [ ] I have updated the
      [change log file](https://github.com/DeepRegNet/DeepReg/blob/main/CHANGELOG.md)
      regarding my changes.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by mathpluscode** [comment]
Testing locally I got

```

Running python demos/unpaired_ct_abdomen/demo_train.py --method weakly --test
2021-02-25 00:52:31.066542: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
WARNING:root:The image loss weight 0.00 is not positive.
2021-02-25 00:52:32.221797: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2021-02-25 00:52:32.221814: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
2021-02-25 00:52:32.221827: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (mathpluscode-XPS-13-9300): /proc/driver/nvidia/version does not exist
2021-02-25 00:52:32.224488: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-02-25 00:52:32.253054: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1497600000 Hz
2021-02-25 00:52:32.253504: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5647a30a8130 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
2021-02-25 00:52:32.253530: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
WARNING:root:The weight for loss image is zero.Loss is not used.
2021-02-25 00:52:36.467187: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session started.
WARNING:tensorflow:Gradients do not exist for variables ['Unet/deconv3d_block_2/conv3d_transpose_2/kernel:0', 'Unet/deconv3d_block_2/batch_normalization_20/gamma:0', 'Unet/deconv3d_block_2/batch_normalization_20/beta:0', 'Unet/deconv3d_block_1/conv3d_transpose_1/kernel:0', 'Unet/deconv3d_block_1/batch_normalization_16/gamma:0', 'Unet/deconv3d_block_1/batch_normalization_16/beta:0', 'Unet/deconv3d_block/conv3d_transpose/kernel:0', 'Unet/deconv3d_block/batch_normalization_12/gamma:0', 'Unet/deconv3d_block/batch_normalization_12/beta:0', 'Unet/sequential_6/conv3d_block_6/conv3d_18/kernel:0', 'Unet/sequential_6/conv3d_block_6/batch_normalization_21/gamma:0', 'Unet/sequential_6/conv3d_block_6/batch_normalization_21/beta:0', 'Unet/sequential_6/residual_conv3d_block_6/conv3d_19/kernel:0', 'Unet/sequential_6/residual_conv3d_block_6/conv3d_20/kernel:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_22/gamma:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_22/beta:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_23/gamma:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_23/beta:0', 'Unet/sequential_5/conv3d_block_5/conv3d_15/kernel:0', 'Unet/sequential_5/conv3d_block_5/batch_normalization_17/gamma:0', 'Unet/sequential_5/conv3d_block_5/batch_normalization_17/beta:0', 'Unet/sequential_5/residual_conv3d_block_5/conv3d_16/kernel:0', 'Unet/sequential_5/residual_conv3d_block_5/conv3d_17/kernel:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_18/gamma:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_18/beta:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_19/gamma:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_19/beta:0', 'Unet/sequential_4/conv3d_block_4/conv3d_12/kernel:0', 'Unet/sequential_4/conv3d_block_4/batch_normalization_13/gamma:0', 'Unet/sequential_4/conv3d_block_4/batch_normalization_13/beta:0', 'Unet/sequential_4/residual_conv3d_block_4/conv3d_13/kernel:0', 'Unet/sequential_4/residual_conv3d_block_4/conv3d_14/kernel:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_14/gamma:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_14/beta:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_15/gamma:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_15/beta:0'] when minimizing the loss.

```

Something must be wrong somewhere.

**Comment by mathpluscode** [comment]
> Testing locally I got
> 
> ```
> 
> Running python demos/unpaired_ct_abdomen/demo_train.py --method weakly --test
> 2021-02-25 00:52:31.066542: I tensorflow/stream_executor/platform/default/dso_loader.cc:48] Successfully opened dynamic library libcudart.so.10.1
> WARNING:root:The image loss weight 0.00 is not positive.
> 2021-02-25 00:52:32.221797: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
> 2021-02-25 00:52:32.221814: W tensorflow/stream_executor/cuda/cuda_driver.cc:312] failed call to cuInit: UNKNOWN ERROR (303)
> 2021-02-25 00:52:32.221827: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (mathpluscode-XPS-13-9300): /proc/driver/nvidia/version does not exist
> 2021-02-25 00:52:32.224488: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN)to use the following CPU instructions in performance-critical operations:  AVX2 AVX512F FMA
> To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
> 2021-02-25 00:52:32.253054: I tensorflow/core/platform/profile_utils/cpu_utils.cc:104] CPU Frequency: 1497600000 Hz
> 2021-02-25 00:52:32.253504: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5647a30a8130 initialized for platform Host (this does not guarantee that XLA will be used). Devices:
> 2021-02-25 00:52:32.253530: I tensorflow/compiler/xla/service/service.cc:176]   StreamExecutor device (0): Host, Default Version
> WARNING:root:The weight for loss image is zero.Loss is not used.
> 2021-02-25 00:52:36.467187: I tensorflow/core/profiler/lib/profiler_session.cc:164] Profiler session started.
> WARNING:tensorflow:Gradients do not exist for variables ['Unet/deconv3d_block_2/conv3d_transpose_2/kernel:0', 'Unet/deconv3d_block_2/batch_normalization_20/gamma:0', 'Unet/deconv3d_block_2/batch_normalization_20/beta:0', 'Unet/deconv3d_block_1/conv3d_transpose_1/kernel:0', 'Unet/deconv3d_block_1/batch_normalization_16/gamma:0', 'Unet/deconv3d_block_1/batch_normalization_16/beta:0', 'Unet/deconv3d_block/conv3d_transpose/kernel:0', 'Unet/deconv3d_block/batch_normalization_12/gamma:0', 'Unet/deconv3d_block/batch_normalization_12/beta:0', 'Unet/sequential_6/conv3d_block_6/conv3d_18/kernel:0', 'Unet/sequential_6/conv3d_block_6/batch_normalization_21/gamma:0', 'Unet/sequential_6/conv3d_block_6/batch_normalization_21/beta:0', 'Unet/sequential_6/residual_conv3d_block_6/conv3d_19/kernel:0', 'Unet/sequential_6/residual_conv3d_block_6/conv3d_20/kernel:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_22/gamma:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_22/beta:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_23/gamma:0', 'Unet/sequential_6/residual_conv3d_block_6/batch_normalization_23/beta:0', 'Unet/sequential_5/conv3d_block_5/conv3d_15/kernel:0', 'Unet/sequential_5/conv3d_block_5/batch_normalization_17/gamma:0', 'Unet/sequential_5/conv3d_block_5/batch_normalization_17/beta:0', 'Unet/sequential_5/residual_conv3d_block_5/conv3d_16/kernel:0', 'Unet/sequential_5/residual_conv3d_block_5/conv3d_17/kernel:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_18/gamma:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_18/beta:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_19/gamma:0', 'Unet/sequential_5/residual_conv3d_block_5/batch_normalization_19/beta:0', 'Unet/sequential_4/conv3d_block_4/conv3d_12/kernel:0', 'Unet/sequential_4/conv3d_block_4/batch_normalization_13/gamma:0', 'Unet/sequential_4/conv3d_block_4/batch_normalization_13/beta:0', 'Unet/sequential_4/residual_conv3d_block_4/conv3d_13/kernel:0', 'Unet/sequential_4/residual_conv3d_block_4/conv3d_14/kernel:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_14/gamma:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_14/beta:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_15/gamma:0', 'Unet/sequential_4/residual_conv3d_block_4/batch_normalization_15/beta:0'] when minimizing the loss.
> ```
> 
> Something must be wrong somewhere.

This was due to a bug in the implementation of UNet, it has been fixed. Therefore irrelevant to this issue.

**Comment by mathpluscode** [comment]
@NMontanaBrown @zacbaum @YipengHu Please have a look as this PR ;) In general, I introduced a new pre-commit hook which checks the type hinting in python. This helps us to better understand the code, and it is also easier to debug ;)

**Review by YipengHu — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -13,6 +13,7 @@ compatible with the updates.
 
 ### Added
 
+- Added support to custom layer channels in U-Net.
```
Comment (zacbaum) [inline CHANGELOG.md:16]: This has nothing to do with the ticket in question... This change really should be it's own ticket and PR... Can you remove this (and any accompanying code) from this PR, and add into this the addition of mypy / updated tests?

File: deepreg/model/backbone/u_net.py [hunk 2]
```diff
@@ -243,10 +253,39 @@ def build_skip_block(self) -> Union[tf.keras.Model, tfkl.Layer]:
         else:
             return tfkl.Add()
 
+    def build_decode_conv_block(
```
Comment (zacbaum) [inline deepreg/model/backbone/u_net.py:256]: As w/ previous comment; this doesn't belong with this PR. This is totally unrelated (unless I'm missing something...)

**Review by zacbaum — CHANGES_REQUESTED** [review]
See comments. I feel like some "extras" have been lumped in with this PR...

File: CHANGELOG.md (hunk 1 above)
Comment (mathpluscode) [reply CHANGELOG.md]: it's gone now, it was due to the fact that this branch was in advance of `main` and at that time there was another branch not merged into `main` yet. ^^

File: deepreg/model/backbone/u_net.py (hunk 2 above)
Comment (mathpluscode) [reply deepreg/model/backbone/u_net.py]: same as previous, it's gone ^^

**Comment by mathpluscode** [comment]
> See comments. I feel like some "extras" have been lumped in with this PR...

Hi @zacbaum thx! The extras should be gone now ;)

**Review by zacbaum — APPROVED** [review]
